##################################################
# RNA-seq data normalization using spike-ins

rm(list=ls())


library("affy")
library("edgeR")



# Load matrix including all samples and RNA-seq reads per gene (generated by htseq-count)
reads_hg38_ERCC=as.data.frame(read.table("count_matrix.all_samples.txt",sep="\t", header=TRUE, row.names=1, dec=".", fill = TRUE))



expressed_hg38_ERCC <- rowSums(reads_hg38_ERCC) > 0
reads_hg38_ERCC <- reads_hg38_ERCC[expressed_hg38_ERCC, ]

reads_hg38_ERCC <- reads_hg38_ERCC+1

dim(reads_hg38_ERCC)
#38194    26 (# genes and ERCCs >0)





# Calculate normalized read counts per million

norm_factors_reads_hg38_ERCC <- calcNormFactors(reads_hg38_ERCC, method = "none")
reads_hg38_ERCC_cpm <- cpm(reads_hg38_ERCC, log = TRUE,
                      lib.size = colSums(reads_hg38_ERCC) * norm_factors_reads_hg38_ERCC)


write.table(reads_hg38_ERCC_cpm, file=paste("MGs_RNAseq_CPM_normalized_matrix.including_ERCCs.txt",sep=""),quote=FALSE,sep="\t",row.names=T,col.names=T)


# Prepare ERCC data
# Obtain the row indices for the ERCC spike-ins and the genes.

ercc_rows <- grep("ERCC", rownames(reads_hg38_ERCC))
gene_rows <- grep("ENSG", rownames(reads_hg38_ERCC))



# Apply LOESS normalization (from the package affy)

reads_hg38_ERCC_loess <- normalize.loess(mat = as.matrix(reads_hg38_ERCC_cpm),
                              subset = ercc_rows, log.it = FALSE)


write.table(reads_hg38_ERCC_loess, file=paste("MGs_RNAseq_CPM_ERCC_loess_normalized_matrix.including_ERCCs.txt",sep=""),quote=FALSE,sep="\t",row.names=T,col.names=T)

